src/swap/switchain.js

```diff
@@ -0,0 +1,371 @@
+// @flow
+
+import { add, div, gt, lt, mul, sub } from 'biggystring'
+import {
+  type EdgeCorePluginOptions,
+  type EdgeCurrencyWallet,
+  type EdgeFetchResponse,
+  type EdgeSpendInfo,
+  type EdgeSpendTarget,
+  type EdgeSwapInfo,
+  type EdgeSwapPlugin,
+  type EdgeSwapPluginQuote,
+  type EdgeSwapRequest,
+  type EdgeTransaction,
+  SwapAboveLimitError,
+  SwapBelowLimitError,
+  SwapCurrencyError
+} from 'edge-core-js/types'
+
+// import { makeSwapPluginQuote } from '../swap-helpers.js'
+
+const swapInfo: EdgeSwapInfo = {
+  pluginId: 'switchain',
+  pluginName: 'switchain',
+  displayName: 'Switchain',
+  quoteUri: 'https://www.switchain.com/transactions',
+  supportEmail: 'help@switchain.com'
+}
+
+let apiUrl = 'https://api.switchain.com/rest/v1'
+
+type SwitchainResponseError = {
+  error: string,
+  reason?: string
+}
+
+type SwitchainOfferResponse = {
+  pair: string,
+  signature: string,
+  quote: string,
+  maxLimit: string,
+  minLimit: string,
+  expiryTs: number,
+  minerFee: string,
+  orderId?: string
+}
+
+type SwitchainOrderCreationResponse = {
+  orderId: string,
+  fromAmount: string,
+  rate: string,
+  exchangeAddress: string,
+  exchangeAddressTag?: string,
+  refundAddress: string,
+  refundAddressTag?: string,
+  toAddres: string,
+  toAddressTag?: string
+}
+
+type SwitchainOrderCreationBody = {
+  pair: string,
+  toAddress: string,
+  toAddressTag?: string,
+  refundAddress: string,
+  refundAddressTag?: string,
+  signature: string,
+  fromAmount?: string,
+  toAmount?: string
+}
+
+const dontUseLegacy = {
+  DGB: true
+}
+
+export function makeSwitchainPlugin(
+  opts: EdgeCorePluginOptions
+): EdgeSwapPlugin {
+  const { initOptions, io } = opts
+
+  if (!initOptions.apiKey) {
+    throw new Error('No Switchain API key provided.')
+  }
+
+  async function getAddress(
+    wallet: EdgeCurrencyWallet,
+    currencyCode: string
+  ): Promise<string> {
+    const addressInfo = await wallet.getReceiveAddress({ currencyCode })
+
+    return addressInfo.legacyAddress && !dontUseLegacy[currencyCode]
+      ? addressInfo.legacyAddress
+      : addressInfo.publicAddress
+  }
+
+  async function swHttpCall(
+    path: string,
+    method: string,
+    body?: Object,
+    query?: { [string]: string }
+  ) {
+    let requestOpts = {
+      headers: {
+        'Content-Type': 'application/json',
+        Authorization: `Bearer ${initOptions.apiKey}`
+      },
+      method
+    }
+
+    if (body) {
+      requestOpts = { ...requestOpts, body: JSON.stringify(body) }
+    }
+    let queryParams = ''
+    if (query) {
+      // only way i found to circumvent Object.entries type errors
+      const queryStringList: Array<[string, string]> = Object.entries(
```

Comment:
Our app needs to run in really old & bad Android WebView implementations that don't have these cool new `Object` methods. I suggest something more like:

```typescript
const queryStringList: Array<[string, string]> = []
for (const k of Object.keys(query)) {
  queryStringList.push([key, query[k]])
}
queryParams = `?${new URLSearchParams(queryStringList).toString()}`
```

Sorry.

Reply:
Thank you for the workaround!